wgpu-core/src/instance.rs

```diff
@@ -652,40 +659,74 @@ impl<G: GlobalIdentityHandlerFactory> Global<G> {
                 desc.extensions - adapter.extensions
             );
 
-            // Check features needed by extensions
-            enabled_features.set(
-                hal::Features::SAMPLER_ANISOTROPY,
-                desc.extensions
-                    .contains(wgt::Extensions::ANISOTROPIC_FILTERING),
-            );
+            // Verify extension preconditions
             if desc
                 .extensions
                 .contains(wgt::Extensions::MAPPABLE_PRIMARY_BUFFERS)
                 && adapter.raw.info.device_type == hal::adapter::DeviceType::DiscreteGpu
             {
                 log::warn!("Extension MAPPABLE_PRIMARY_BUFFERS enabled on a discrete gpu. This is a massive performance footgun and likely not what you wanted");
             }
+
+            let available_features = adapter.raw.physical_device.features();
+
+            // Check features that are always needed
+            let wishful_features = hal::Features::VERTEX_STORES_AND_ATOMICS
+                | hal::Features::FRAGMENT_STORES_AND_ATOMICS
+                | hal::Features::NDC_Y_UP;
+            let mut enabled_features = available_features & wishful_features;
+            if enabled_features != wishful_features {
+                log::warn!(
+                    "Missing features: {:?}",
+                    wishful_features - enabled_features
+                );
+            }
+
+            // Extensions
             enabled_features.set(
-                hal::Features::TEXTURE_DESCRIPTOR_ARRAY,
-                desc.extensions
-                    .contains(wgt::Extensions::SAMPLED_TEXTURE_BINDING_ARRAY),
-            );
-            enabled_features.set(
-                hal::Features::SHADER_SAMPLED_IMAGE_ARRAY_DYNAMIC_INDEXING,
+                hal::Features::SAMPLER_ANISOTROPY,
                 desc.extensions
-                    .contains(wgt::Extensions::SAMPLED_TEXTURE_ARRAY_DYNAMIC_INDEXING),
+                    .contains(wgt::Extensions::ANISOTROPIC_FILTERING),
             );
+
+            // If extensions are not enabled, we must disable their capabilities
```

Comment:
could we maybe refactor this a bit: instead of removing the capabilities from whatever is stored in the adapter, populate them according to the enabled extensions?